presto-spi/src/main/java/com/facebook/presto/spi/connector/DelegateConnectorAccessControl.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.connector;
+
+import com.facebook.presto.spi.SchemaTableName;
+import com.facebook.presto.spi.security.Identity;
+import com.facebook.presto.spi.security.Privilege;
+
+import java.util.Set;
+import java.util.function.Supplier;
+
+import static java.util.Objects.requireNonNull;
+
+public class DelegateConnectorAccessControl
+        implements ConnectorAccessControl
+{
+    private final Supplier<ConnectorAccessControl> delegate;
```

Comment:
I find Guava's pattern of `protected abstract ... delegate();` (eg `com.google.common.collect.ForwardingMap`) much more flexible and easier to use.

Reply:
I think it is because you get used to it. Notice that when one implements `ForwardingMap#delegate` it does logically exactly the same as providing `Supplier` here, they inject a method implementation. However here is better due usage of composition, not inheritance.

> much more flexible and easier to use.

Can you please point me an example where this approach is worse?

Reply:
It's not better.
Have you ever faced the disastrous inflexibility `AbstractTestQueryFramework` imposes on you, as it accepts `QueryRunnerSupplier` which you need to pass in `super` before you had a chance to prepare it?

I feel strongly about it.